webserver/templates/user/scraper_api.js

```diff
@@ -0,0 +1,309 @@
+function makeURIObject(lastfmURI, spotifyURI) {
+
+}
+
+function select(selector, collection) {
+    var newCollection = [];
+    for (var i = 0; i < collection.length; i++) {
+        if (selector(collection[i])) {
+            newCollection.push(collection[i]);
+        }
+    }
+
+    return newCollection;
+}
+
+function map(applicable, collection) {
+    var newCollection = [];
+    for (var i = 0; i < collection.length; i++) {
+        newCollection.push(applicable(collection[i]));
+    }
+
+    return newCollection;
+}
+
+function each(applicable, collection) {
+    for (var i = 0; i < collection.length; i++) {
+        applicable(collection[i]);
+    }
+}
+
+function isSpotifyURI(uri) {
+    return !!(/open.spotify/.exec(uri));
+}
+
+function Scrobble(rootScrobbleElement) {
+    this.rootScrobbleElement = rootScrobbleElement;
+}
+
+Scrobble.prototype.lastfmID = function () {
+    var loveButtonURL = this.rootScrobbleElement['url'];
+    return extractlastFMIDFromURL(loveButtonURL);
+};
+
+Scrobble.prototype.artistName = function () {
+    var artistName = this.rootScrobbleElement['artist']['#text'];
+    return artistName;
+};
+
+Scrobble.prototype.trackName = function () {
+    var trackData = this.rootScrobbleElement['name'];
+    return trackData;
+};
+
+Scrobble.prototype.scrobbledAt = function () {
+    var date = this.rootScrobbleElement['date']['uts'];
+    return date;
+};
+
+Scrobble.prototype.optionalSpotifyID = function () {
+    return select(
+            isSpotifyURI,
+            map(
+                function(elem) { return elem.getAttribute("href") },
+                this.rootScrobbleElement.getElementsByTagName("a")
+               )
+            )[0];
+};
+
+Scrobble.prototype.asJSONSerializable = function () {
+    return {
+        "track_metadata": {
+            "track_name": this.trackName(),
+            "artist_name": this.artistName(),
+        },
+        "listened_at": this.scrobbledAt()
+
+    }
+};
+
+function extractlastFMIDFromURL(loveButtonURL) {
+    var parts = loveButtonURL.split("/");
+    return parts.slice(0, parts.length-2).join("/");
+}
+
+function encodeScrobbles(jsonstr) {
+    var scrobbles = JSON.parse(jsonstr);
+    scrobbles = scrobbles['recenttracks']['track'];
```

Comment:
the API includes much richer data - in particular, it will include a recording MBID if the user's client submitted it. We should include that here